allow fetching dashboard info for other users

aliceriot opened the pull request:
#### What are the relevant tickets?

this closes #2531

#### What's this PR do?

This does just the backend work to support fetching dashboard information based on a username URL parameter. For now the only change in the frontend it to the `getDashboard` function and adding `SETTINGS.user.username` in when necessary.

#### How should this be manually tested?

Make sure that the dashboard and all the frontend stuff works fine. If it does, I think things should be good.

If you wanted you could do the following (but I think it's not essential):

1. edit `DashboardPage.js` or `App.js`, adding:

```js
import * as api from '../lib/api';

window.api = api
```

then, from the JavaScript console, call:

```js
api.getDashboard('fake_username');
```

this should return a 404. confirm getting your own username returns a 200.

then, find the username of a real learner in your database, and ensure your user does not have a `Staff` or `Instructor` role in a program that learner is enrolled in. then call:

```js
api.getDashboard('real_learner_username');
```

this should return a 404. then add a staff or instructor role to your user for that program, and try again. you should now get a 200 and the user's dashboard data.

noisecapella commented:
I investigated why the selenium test was failing and it looks like it's because you're using `User.username` but the client is passing the edX username. Can you change the view to look up the username in the social auth model?

noisecapella commented on dashboard/permissions.py:
```diff
@@ -0,0 +1,39 @@
+        if request.user.is_anonymous():
+            raise Http404
+
+        user = get_object_or_404(User, username=view.kwargs['username'])
```
We should look up the user by the social auth username here too

aliceriot replied:
Hm, I think I was under the impression that `user.username` was the same thing. How would I go about doing that? If someone visits `/profile/${username}` is `username` guaranteed to be someone's edx username?

noisecapella replied:
Yes, the only username we expose to the outside user is their edX username. Usually this will be the same as `User.username` but if there's a conflict or an error when creating the social auth object it may create a separate user object with some hex digits appended to the username, ie `george6fb83a` for an edX username of `george`. You can look it up by joining the social auth table like this: `User.objects.filter(social_auth__uid='a_username_here')`

aliceriot replied:
OK, I'll make that change.

noisecapella replied:
Actually you should also filter on the provider:

    User.objects.filter(social_auth_uid='a_username_here', social_auth__provider=EdxOrgOAuth2.name)

aliceriot replied:
makes sense!

noisecapella commented on dashboard/permissions_test.py:
```diff
@@ -0,0 +1,96 @@
+            with self.assertRaises(Http404):
+                perm.has_permission(request, view)
+
+    @ddt.data((Instructor), (Staff))
```
Can you remove the extra parentheses here?

aliceriot replied:
huh, before I had to add those because of some error, but now it's not happening! lol

noisecapella commented on dashboard/permissions_test.py:
```diff
@@ -0,0 +1,96 @@
+    def test_staff_can_read_learners(self, role):
+        """
+        A staff or instructor on a program a learner is also in should
+        be able to read their profile
```
Can you change to `read their dashboard`?

aliceriot replied:
oop good catch

noisecapella commented on dashboard/views_test.py:
```diff
@@ -80,6 +87,23 @@ def test_get_dashboard(self):
         assert len(result.data) == 2
         assert [self.program_1.id, self.program_2.id] == [res_item['id'] for res_item in result.data]
 
+    @ddt.data((Instructor), (Staff))
```
Can you remove the extra parentheses here?

noisecapella commented on static/js/actions/index.js:
```diff
@@ -35,10 +36,10 @@ export const receiveDashboardFailure = (errorInfo: APIErrorInfo): Action => ({
     dispatch(requestDashboard(noSpinner));
-    return api.getDashboard().
+    return api.getDashboard(username).
       then(dashboard => dispatch(receiveDashboardSuccess(dashboard))).
```
We need to update the reducer to handle multiple dashboards, else dashboards from other users will overwrite each other

aliceriot replied:
yes, I'm planning to do that work in a separate PR to keep the changeset small.

aliceriot replied:
Provided that makes sense to you. All I've changed is on the backend we can now theoretically get the information for another user, but on the frontend all the calls to `getDashboard` use the user's own username. The next PR I do will change the reducer to support multiple dashboards (similar to the setup we have for profiles) and then after that one I will probably work on actually adding the dashboard support to the learner page (along with the UI work, some of which I've already started).

noisecapella replied:
Sounds good to me

aliceriot commented:
@noisecapella I've changed things around to fetch users based on the edx social auth stuff (and refactored some tests as well). ready for review again!
